File: MCRRTImp/src/test/java/infrastructure/model/Attacker.java
package infrastructure.model;

import infrastructure.Vector2BasedImp.Vector2;
import lab.mars.MCRRTImp.algorithm.MCRRT;
import lab.mars.MCRRTImp.algorithm.MCTSSampler;
import lab.mars.MCRRTImp.model.*;
import lab.mars.RRTBase.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedDeque;
import java.util.concurrent.ConcurrentLinkedQueue;

public class Attacker<V extends Vector<V>> extends SimulatedVehicle<V> {

    private double viewDistance;

    private double viewAngle;

    private double topVelocity;

    private DimensionalWayPoint<V> designatedTargetPosition;

    private Queue<DimensionalPath<DimensionalWayPoint<V>>> actualPath = new ConcurrentLinkedQueue<>();

    private Queue<DimensionalPath<DimensionalWayPoint<V>>> areaPath = new ConcurrentLinkedQueue<>();

    private ScaledGrid<V> gridWorld;

    private RRT algorithm;

    private Provider<List<Obstacle<V>>> obstacleProvider;

    private List<NTreeNode<DimensionalWayPoint<V>>> leaves = new ArrayList<>();

    public List<NTreeNode<DimensionalWayPoint<V>>> getLeaves() {
        return leaves;
    }

    private void leafApplier(List<NTreeNode<DimensionalWayPoint<V>>> leaves) {
        this.leaves = leaves;
    }


    public Attacker(V position, V velocity, double rotationLimits, int numberOfDirection, double safeDistance, double viewDistance, double viewAngle, DimensionalWayPoint<V> designatedTargetPosition,
                    Space<V> area, Provider<List<Obstacle<V>>> obstacleProvider) {
        super(position, velocity, rotationLimits, numberOfDirection, safeDistance);
        this.obstacleProvider = obstacleProvider;
        this.topVelocity = velocity.len();
        this.viewDistance = viewDistance;
        this.designatedTargetPosition = designatedTargetPosition;
        this.viewAngle = viewAngle;
        this.algorithm = new MCRRT<>(1 / 2.0, area, null, () -> 1000,   obstacleProvider, () -> this, () -> designatedTargetPosition, (set) -> {},  this::setActualPath, false);
//        this.algorithm = new MCTSSampler<>(10, area, obstacleProvider, () -> this, () -> designatedTargetPosition, this::setActualPath, null, this::leafApplier, null);

    }

    public void setDesignatedTarget(DimensionalWayPoint<V> target) {
        this.designatedTargetPosition = target;
    }

    public DimensionalPath<DimensionalWayPoint<V>> actualPath() {
        while (actualPath.size() > 1) {
            actualPath.poll();
        }
        return actualPath.peek();
    }

    public DimensionalWayPoint<V> target() {
        return designatedTargetPosition;
    }

    public ScaledGrid<V> gridWorld() {
        return gridWorld;
    }

    public void setAreaPath(DimensionalPath<DimensionalWayPoint<V>> areaPath) {
        this.areaPath.offer(areaPath);
    }

    public void setGridWorld(ScaledGrid<V> gridWorld) {
        this.gridWorld = gridWorld;
    }

    public void setActualPath(DimensionalPath<DimensionalWayPoint<V>> actualPath) {
        this.actualPath.offer(actualPath);
//        startAlgorithm();
    }

    public Queue<DimensionalPath<DimensionalWayPoint<V>>> areaPath() {
        return this.areaPath;
    }

    @Override
    protected double simulateVelocity(double currentVelocity, double angle) {
        return topVelocity * (1 - Math.abs(angle) / this.rotationLimits);
    }

    public double viewDistance() {
        return viewDistance;
    }

    public double viewAngle() {
        return viewAngle;
    }


    public void startAlgorithm() {
        new Thread(() -> {
            while (true) {
                algorithm.solve(true);
                V cpy = position.cpy();
                boolean flag = false;
                do {
                    cpy.dimensions()[0].value = MathUtil.random(0.0, 1920.0);
                    cpy.dimensions()[1].value = MathUtil.random(0.0, 1080.0);
                    flag = false;
                    for (Obstacle<V> obstacle : obstacleProvider.provide()) {
                        if (obstacle.contains(cpy)) {
                            flag = true;
                        }
                    }
                }while (flag);
                position.set(cpy);
            }
        }).start();
    }

}

File: MCRRTImp/src/test/kotlin/infrastructure/model/Attacker.kt
package infrastructure.model

import lab.mars.MCRRTImp.algorithm.MCRRT
import lab.mars.MCRRTImp.base.*
import lab.mars.MCRRTImp.model.*

import java.util.ArrayList
import java.util.Queue
import java.util.concurrent.ConcurrentLinkedQueue

class Attacker<V : Vector<V>>(position: V,
                              velocity: V,
                              rotationLimits: Double,
                              numberOfDirection: Int,
                              safeDistance: Double,
                              private val viewDistance: Double,
                              private val viewAngle: Double,
                              private val designatedTargetPosition: WayPoint<V>?,
                              area: Space<V>?,
                              private val obstacleProvider: (() -> List<Obstacle<V>>)?) : SimulatedVehicle<V>(position, velocity, rotationLimits, numberOfDirection, safeDistance) {

    private val topVelocity: Double = velocity.len()

    private val actualPath = ConcurrentLinkedQueue<Path<WayPoint<V>>>()

    private val areaPath = ConcurrentLinkedQueue<Path<WayPoint<V>>>()

    private var gridWorld: DiscreteWorld<V>? = null

    private lateinit var algorithm: MCRRT<V>

    init {
        if (area != null && obstacleProvider != null && designatedTargetPosition != null)
            algorithm = MCRRT(
                    spaceRestriction = area,
                    obstacleProvider = obstacleProvider,
                    vehicleProvider = { this },
                    targetProvider = { designatedTargetPosition },
                    verbose = false,
                    firstLevelDeltaTime = 600.0,
                    secondLevelDeltaTime = 10.0
            )
    }

    var leaves: List<NTreeNode<WayPoint<V>>> = ArrayList()
        private set

    private fun leafApplier(leaves: List<NTreeNode<WayPoint<V>>>) {
        this.leaves = leaves
    }


    fun setDesignatedTarget(target: WayPoint<V>) {
        this.designatedTargetPosition?.apply {
            origin.set(target.origin)
            velocity.set(target.velocity)
        }
    }

    fun actualPath(): List<Path<WayPoint<V>>> {
        val ret = ArrayList<Path<WayPoint<V>>>()
        actualPath.forEach { ret.add(it) }
        ret.sortBy {  it.utility }
        return ret
    }

    fun target(): WayPoint<V>? {
        return designatedTargetPosition
    }

    fun gridWorld(): DiscreteWorld<V>? {
        return gridWorld
    }

    fun setAreaPath(areaPath: Path<WayPoint<V>>) {
        this.areaPath.offer(areaPath)
    }

    fun setGridWorld(gridWorld: DiscreteWorld<V>) {
        this.gridWorld = gridWorld
    }

    fun setActualPath(actualPath: Path<WayPoint<V>>) {
        this.actualPath.offer(actualPath)
        //        startAlgorithm();
    }

    fun areaPath(): Queue<Path<WayPoint<V>>> {
            return this.areaPath
    }

    override fun simulateVelocity(currentVelocity: Double, angle: Double): Double {
        return topVelocity * (1 - Math.abs(angle) / this.rotationLimits)
    }

    fun viewDistance(): Double {
        return viewDistance
    }

    fun viewAngle(): Double {
        return viewAngle
    }


    fun startAlgorithm() {

        Thread {
            var i = 0
            while (true) {
//                i++
                println(i)
                val configuration : OneTimeConfiguration<V>
                if (i > 1) {
                    configuration = OneTimeConfiguration(
                            timeTolerance = Long.MAX_VALUE,
                            levelOneReplan = true,
                            levelTwoReplan = false,
                            wayPointApproachDistance = 10.0,
                            levelOneUseCache = true,
                            levelOnePathCache = actualPath.peek(),
                            levelOneReplanCacheIdx = 10

                    )
                } else {
                    configuration = OneTimeConfiguration(
                            timeTolerance = Long.MAX_VALUE,
                            levelOneReplan = true,
                            levelTwoReplan = true,
                            levelTwoFromIdx = 0,
                            levelTwoToIdx = 100,
                            wayPointApproachDistance = 10.0,
                            levelOneUseCache = false)
                }
                algorithm.solve(configuration) { value ->
                    when (value.status) {
                        MCRRT.ResultStatus.InProgress, MCRRT.ResultStatus.Complete -> {
                            actualPath.clear()
                            actualPath.add(value.areaPath!!)
                        }
                        else -> {
                            println(value.status)
                        }
                    }
                }
                val cpy = position.cpy()
                var flag: Boolean
                do {
                    cpy.dimensions[0].value = 0.0 random 1920.0
                    cpy.dimensions[1].value = 0.0 random 1080.0
                    flag = false
                    if (obstacleProvider != null) {
                        for (obstacle in obstacleProvider.invoke()) {
                            if (obstacle.contains(cpy)) {
                                flag = true
                            }
                        }
                    }
                } while (flag)
                position.set(cpy)
            }
            println("over")
        }.start()
    }

}
